### PR TITLE
esptool: Allow to pass the partition table CSV.

### PR DESCRIPTION
**Pull request by iosabi**
### Contribution description

The partition table of the device in the esp8266 and esp32 based boards
was set to a default table with one "factory" partition with exactly
the size of the compiled firmware. This is problematic if we want to
update the device on the field.

This patch allows to set the `PARTITION_TABLE_CSV` variable from the
Makefile to a .csv file with a custom partition table, for example this
could be set to a partition table with two ota entries, or with a single
factory entry but of a known fixed size.

As a side effect of the make cleanup in this patch we now support
passing `-j` to the `make flash` command so we can compile in parallel
and still run the flash commands only once at the end. Before this
patch the conversion from .elf to .elf.bin was happening before the
code was recompiled when running in parallel.

### Testing procedure

`make BOARD=esp8266-esp-12x Q= -C tests/lwip flash -j term`

Verified that the compile commands at the end (after the link step) appear in the right order.

`make BOARD=esp8266-esp-12x PARTITION_TABLE_CSV=mytable.csv -C tests/lwip flash -j term`

Verified that the `mytable.csv` file is used instead.

### Issues/PRs references

None

### CODE REVIEW

**Comment by iosabi** [comment]
@gschorcht FYI

File: makefiles/tools/esptool.inc.mk [hunk 1]
```diff
@@ -10,10 +10,11 @@ BOOTLOADER_BIN = bootloader$(BOOTLOADER_COLOR)$(BOOTLOADER_INFO).bin
 ESPTOOL ?= $(RIOTTOOLS)/esptool/esptool.py
 
 # The ELFFILE is the base one used for flashing
+ELFFILE ?= $(BINDIR)/$(APPLICATION).elf
```
Comment (fjmolinas) [inline makefiles/tools/esptool.inc.mk:13]: Why does this need defining?

Comment (fjmolinas) [reply makefiles/tools/esptool.inc.mk:13]: I mean shouldn't it be known?

**Review by fjmolinas — COMMENTED** [review]
Seems sensible, but I dont have a BOARD to test, maybe @benpicco or @gschorcht could give it a run?

**Comment by fjmolinas** [comment]
@kaspar030 is the esp32-wroom32 still on the ci?

**Comment by fjmolinas** [comment]
> Seems sensible, but I dont have a BOARD to test, maybe @benpicco or @gschorcht could give it a run?

I triggered the CI to run rests on that BOARD, it should tell  us if its OK.

File: makefiles/tools/esptool.inc.mk (hunk 1 above)
Comment (iosabi) [reply makefiles/tools/esptool.inc.mk:13]: Hi, it is not know at this stage.
While not having `ELFFILE` set at this point is fine if you use $(ELFFILE) in the list of commands for a rule, for example the code I'm removing had a `ls -l $(FLASHFILE).bin` command which is fine because at the time those commands are executed the `ELFFILE` variable is already set, it didn't work here because now I'm setting new rules (`$(FLASHFILE).bin: $(FLASHFILE)`) that use `ELFFILE` or `FLASHFILE` in the rule target name and dependencies, so `FLASHFILE` (and `ELFFILE`) need to be evaluated at the time this file is imported. The issue with not listing the `$(FLASHFILE)` as a dependency in the rules is that  `make -j9 flash` doesn't work as expected since the rule is executed too early.

This makefile file is imported when `PROGRAMMER` is set to `esptool`, from here: https://github.com/RIOT-OS/RIOT/blob/fed2c8e87ec0269e53388ad60c030b4f0495a266/Makefile.include#L420
but the `ELFFILE` is set in the Mafile.include a little later, from this line:
https://github.com/RIOT-OS/RIOT/blob/fed2c8e87ec0269e53388ad60c030b4f0495a266/Makefile.include#L559
so at this point `ELFFILE` is not set.

I don't think this is a problem since the `esptool` programmer is really only useful for the ESP CPUs, so having the default assumption that we are flashing an ELFFILE with this name is fine. This is also a `?=` assignment so the board or app can override `ELFFILE` if they want to.

**Comment by iosabi** [comment]
I don't have a esp32-wrom board to test this, but here's what I did (PROG_DEV=/dev/null because I don't have such device):

```bash
git checkout esp_table^  # Checks out the parent commit
make -C tests/od/ BOARD=esp32-wroom-32 PROG_DEV=/dev/null RIOT_VERSION=esp32-test flash
mv tests/od/bin/esp32-wroom-32 tests/od/bin/esp32-wroom-32-head
```

Note that in the `make` command using `-j` causes this error: `FileNotFoundError: [Errno 2] No such file or directory` on the `test_od.elf` because of the missing dep this commit is also fixing, so no `-j` there.

Then, for this commit:
```bash
git checkout esp_table
make -C tests/od/ BOARD=esp32-wroom-32 PROG_DEV=/dev/null RIOT_VERSION=esp32-test flash -j
mv tests/od/bin/esp32-wroom-32 tests/od/bin/esp32-wroom-32-table
```

Then compare the two directories with `meld`, they are identical. I'm passing RIOT_VERSION= because otherwise the builds would be different due to the different version string.

Looking at the failure and the `.murdock` file, it seems that murdock runs `flash-only` which with this commit is also building the files (because of the dep) but before it wasn't, and the CI fails with ```make[1]: xtensa-esp32-elf-gcc: Command not found```
I'll look into this a bit more.

File: makefiles/tools/esptool.inc.mk [hunk 2]
```diff
@@ -10,10 +10,14 @@ BOOTLOADER_BIN = bootloader$(BOOTLOADER_COLOR)$(BOOTLOADER_INFO).bin
...
 FLASHFILE ?= $(ELFFILE)
 
-# Convert .elf to .bin before flashing
-FLASHDEPS += esp-image-convert
+# Convert .elf and .csv to .bin files at build time. These are needed for
+# flashing only, but if they are added to FLASHDEPS instead a "flash-only"
+# target would cause a rebuild of $(ELFFILE) since these files depend on it and
+# ELFFILE depends on FORCE.
+BUILD_FILES += $(FLASHFILE).bin $(BINDIR)/partitions.bin
```
Comment (iosabi) [inline makefiles/tools/esptool.inc.mk:20]: I changed these to use BUILD_FILES instead of FLASHDEPS even if they are only needed for flashing. The real problem here is that ELFFILE depends on FORCE so we can't have FLASHDEPS depend on the ELFFILE (FLASHFILE) and not trigger a rebuild. I think ELFFILE shouldn't depend on FORCE but instead have some build config dependency on a lazysponge output or similar... but that's for another bug.

Comment (iosabi) [reply makefiles/tools/esptool.inc.mk]: I changed this back. Turns out this doesn't work either because the CI doesn't copy those files.

I don't see a way to set this up correctly because ELFFILE is FORCEd (not sure why). I filed #16385  to address that.

Meanwhile I implemented a somewhat hacky fix setting these as .PHONY when using flash-only. This should make the CI happy.

Comment (kaspar030) [reply makefiles/tools/esptool.inc.mk]: > Turns out this doesn't work either because the CI doesn't copy those files.

you mean when running the tests? Try adding needed files to `TEST_EXTRA_FILES`.

**Comment by fjmolinas** [comment]
@iosabi during a review process, please avoid push forcing unless asked, approved by the reviewer, it makes following your changes harder, especially if you rebase as well, you can take a look at our contributing guidelines for more [add-fixup-commits-during-review](https://github.com/RIOT-OS/RIOT/blob/master/CONTRIBUTING.md#add-fixup-commits-during-review).

File: makefiles/tools/esptool.inc.mk (hunk 1 above)
Comment (fjmolinas) [reply makefiles/tools/esptool.inc.mk:13]: Ok that makes sense, in that case I think that we can add it there, although it seems moving the PROGRAMMER include later is the best fix but is a huge change that would require more testing. @iosabi can you add a comment summarizing what you just described just before `ELFFILE ?= $(BINDIR)/$(APPLICATION).elf`?

File: makefiles/tools/esptool.inc.mk (hunk 2 above)
Comment (fjmolinas) [reply makefiles/tools/esptool.inc.mk]: @iosabi I took a look and I think adding to `BUILD_FILES` is the way to go. Simply add also a line with

```
TEST_EXTRA_FILES += $(FLASHFILE).bin $(BINDIR)/partitions.bin
```

Just below when you add to `BUILD_FILES`

File: makefiles/tools/esptool.inc.mk [hunk 3]
```diff
@@ -10,10 +10,11 @@ BOOTLOADER_BIN = bootloader$(BOOTLOADER_COLOR)$(BOOTLOADER_INFO).bin
 ESPTOOL ?= $(RIOTTOOLS)/esptool/esptool.py
 
 # The ELFFILE is the base one used for flashing
+ELFFILE ?= $(BINDIR)/$(APPLICATION).elf
 FLASHFILE ?= $(ELFFILE)
 
-# Convert .elf to .bin before flashing
-FLASHDEPS += esp-image-convert
+# Convert .elf and .csv to .bin files before flashing.
+FLASHDEPS += $(FLASHFILE).bin $(BINDIR)/partitions.bin
```
Comment (fjmolinas) [inline makefiles/tools/esptool.inc.mk:17]:
```suggestion
BUILD_FILE += $(FLASHFILE).bin $(BINDIR)/partitions.bin
TEST_EXTRA_FILES += $(FLASHFILE).bin $(BINDIR)/partitions.bin
```

File: makefiles/tools/esptool.inc.mk [hunk 4]
```diff
@@ -31,19 +32,38 @@ else
...
 endif
 
-.PHONY: esp-image-convert esp-qemu
-# prepare image to flash: convert .elf to .bin
-esp-image-convert:
+# This target actually depends on $(FLASHFILE), but if we add $(FLASHFILE) here
+# it will cause it to be rebuilt even in "flash-only" mode because ELFFILE is
+# FORCEd. To avoid that we either add the dependency when calling make with
+# "all" or "flash", or we make it a PHONY target when only doing "flash-only".
+$(FLASHFILE).bin:
```
Comment (fjmolinas) [inline makefiles/tools/esptool.inc.mk:39]:
```suggestion
$(FLASHFILE).bin: $(FLASHFILE)
```

Or how about sticking to `ELFFILE` instead?

```suggestion
$(ELFFILE).bin: $(ELFFILE)
```

File: makefiles/tools/esptool.inc.mk [hunk 5]
```diff
@@ -31,19 +32,38 @@ else
...
 	$(Q)$(ESPTOOL) --chip $(FLASH_CHIP) elf2image --flash_mode $(FLASH_MODE) \
 		--flash_size $(FLASH_SIZE)MB --flash_freq $(FLASH_FREQ) $(FLASH_OPTS) \
-		-o $(FLASHFILE).bin $(FLASHFILE)
+		-o $@ $(FLASHFILE)
```
Comment (fjmolinas) [inline makefiles/tools/esptool.inc.mk:42]:
```suggestion
		-o $@ $<
```

File: makefiles/tools/esptool.inc.mk [hunk 6]
```diff
@@ -31,19 +32,38 @@ else
...
+
+ifneq (,$(filter all flash, $(MAKECMDGOALS)))
+$(FLASHFILE).bin: $(FLASHFILE)
+else
+ifneq (,$(filter flash-only, $(MAKECMDGOALS)))
+.PHONY: $(FLASHFILE).bin
+endif
+endif
```
Comment (fjmolinas) [inline makefiles/tools/esptool.inc.mk:50]:
```suggestion
ifneq (,$(filter all flash, $(MAKECMDGOALS)))
$(FLASHFILE).bin: $(FLASHFILE)
else
ifneq (,$(filter flash-only, $(MAKECMDGOALS)))
.PHONY: $(FLASHFILE).bin
endif
endif
```

With all the above changes this would not be needed.

**Review by fjmolinas — COMMENTED** [review]
I took a closer look and I think your initial approach was the correct one @iosabi, just add `TEST_EXTRA_FILES` as suggested by @kaspar030.

Also I think these changes would fix https://github.com/RIOT-OS/RIOT/issues/13492? no?

**Comment by fjmolinas** [comment]
> I took a closer look and I think your initial approach was the correct one @iosabi, just add `TEST_EXTRA_FILES` as suggested by @kaspar030.
> 
> Also I think these changes would fix #13492? no?

I have an `esp32-wroom-32` so I believe I can test this as well, I modified your test procedure to account for the corner cases. Thanks a lot for investigating this!

File: makefiles/tools/esptool.inc.mk (hunk 1 above)
Comment (iosabi) [reply makefiles/tools/esptool.inc.mk:13]: I'll add a comment to that line.

File: makefiles/tools/esptool.inc.mk (hunk 4 above)
Comment (iosabi) [reply makefiles/tools/esptool.inc.mk]: Strictly speaking, I think `FLASHFILE` should be `$(ELFFILE).bin`. I made that change in a second fixup commit now.

File: makefiles/tools/esptool.inc.mk (hunk 5 above)
Comment (iosabi) [reply makefiles/tools/esptool.inc.mk]: Done.

File: makefiles/tools/esptool.inc.mk (hunk 6 above)
Comment (iosabi) [reply makefiles/tools/esptool.inc.mk]: Done.

File: makefiles/tools/esptool.inc.mk (hunk 3 above)
Comment (iosabi) [reply makefiles/tools/esptool.inc.mk]: Done in the first fixup commit.

**Comment by iosabi** [comment]
> I have an `esp32-wroom-32` so I believe I can test this as well, I modified your test procedure to account for the corner cases. Thanks a lot for investigating this!

The last branch should do it. I added two fixup commits. Sorry to force push them earlier, you should be able to see the diff if you click on the [force-pushed](https://github.com/RIOT-OS/RIOT/compare/f7e63ab0cf7f7676600cb6642270c70df9576655..2a8937dbf4ae50935333b38dfa49c41fcb100a67) word in the automatic message from github if you want to see that.

File: makefiles/tools/esptool.inc.mk [hunk 7]
```diff
@@ -28,22 +35,30 @@ else
...
 	$(Q)$(ESPTOOL) --chip $(FLASH_CHIP) elf2image --flash_mode $(FLASH_MODE) \
 		--flash_size $(FLASH_SIZE)MB --flash_freq $(FLASH_FREQ) $(FLASH_OPTS) \
-		-o $(FLASHFILE).bin $(FLASHFILE)
+		-o $@ $<
+
+# Default partition table with no OTA. Can be replaced with a custom partition
+# table setting PARTITION_TABLE_CSV.
+PARTITION_TABLE_CSV ?= $(BINDIR)/partitions.csv
+
+$(BINDIR)/partitions.csv: $(FLASHFILE)
```
Comment (fjmolinas) [inline makefiles/tools/esptool.inc.mk:51]:
```suggestion
$(BINDIR)/partitions.csv: $(ELFFILE).bin
```

**Comment by fjmolinas** [comment]
Looks good @iosabi can you address my final comment, you can squash afterwards!

File: makefiles/tools/esptool.inc.mk (hunk 2 above)
Comment (iosabi) [reply makefiles/tools/esptool.inc.mk]: Adding them to `TEST_EXTRA_FILES` works in the CI but the following doesn't work:

```make -C tests/od BOARD=esp32-wroom-32 all flash-only -j```

however, the following works:

```make -C tests/od BOARD=esp32-wroom-32 flash -j```

which I guess is good enough. I believe that `flash -j` works because `flash: all` dependency which `flash-only` doesn't have, and `all` (well `link`) eventually depends on the `$(BUILD_FILES)`.

File: makefiles/tools/esptool.inc.mk (hunk 7 above)
Comment (iosabi) [reply makefiles/tools/esptool.inc.mk:51]: I believe this one is better to leave as FLASHFILE. If you are using the default partition table but want to use a different FLASHFILE you can do something like:

```make
FLASHFILE = $(BINDIR)/myflashfile.bin
$(BINDIR)/myflashfile.bin: $(ELFFILE).bin my-extra-static-data
  cat $^ > $@
```

The FFLAGS has `0x10000 $(FLASHFILE)` so the default partition table should match that. If we were to implement OTA at some point then we wouldn't be using this default partition table and this point doesn't matter; but I think this should match what FFLAGS has and using FLASHFILE gives a tiny bit more flexibility.

I changed the body of the rule to use `$<` to avoid writing this twice.

**Comment by iosabi** [comment]
> Looks good @iosabi can you address my final comment, you can squash afterwards!

I added the last change to https://github.com/RIOT-OS/RIOT/pull/16307/commits/6053f815f54c7dfe9177b4fb5e8b408a074ffc8d and then squashed and force-pushed.

File: makefiles/tools/esptool.inc.mk (hunk 7 above)
Comment (fjmolinas) [reply makefiles/tools/esptool.inc.mk:51]: Sure

**Comment by fjmolinas** [comment]
@iosabi can you post some kind of results for your custom partition table test?

**Comment by iosabi** [comment]
> @iosabi can you post some kind of results for your custom partition table test?

Sure!

First we need a custom partition table like this:
```
$ cat mypartitions.csv 
# Name,   Type, SubType, Offset,   Size
nvs,      data, nvs,     0x9000,   0x6000
phy_init, data, phy,     0xf000,   0x1000
ota_0,    app,  ota_0,   0x010000, 0x70000
ota_1,    app,  ota_1,   0x080000, 0x70000
otadata,  data, ota,     0x0f0000, 0x2000
```

Then we can run the following command:
```
USEMODULE=esp_log_startup make -C tests/shell BOARD=esp8266-esp-12x PARTITION_TABLE_CSV=`pwd`/mypartitions.csv -j flash
```


```
I (66) boot: SPI Mode       : DOUT
I (69) boot: SPI Flash Size : 4MB
I (72) boot: Partition Table:
I (74) boot: ## Label            Usage          Type ST Offset   Length
I (81) boot:  0 nvs              WiFi data        01 02 00009000 00006000
I (87) boot:  1 phy_init         RF data          01 01 0000f000 00001000
I (93) boot:  2 ota_0            OTA app          00 10 00010000 00070000
I (100) boot:  3 ota_1            OTA app          00 11 00080000 00070000
I (106) boot:  4 otadata          OTA data         01 00 000f0000 00002000
I (113) boot: End of partition table
I (116) boot: No factory image, trying OTA 0
I (120) esp_image: segment 0: paddr=0x00010010 vaddr=0x40210010 size=0x3b950 (244048) map
I (250) esp_image: segment 1: paddr=0x0004b968 vaddr=0x3ffe8000 size=0x0112c (  4396) load
I (252) esp_image: segment 2: paddr=0x0004ca9c vaddr=0x3ffe912c size=0x0039c (   924) load
I (255) esp_image: segment 3: paddr=0x0004ce40 vaddr=0x40100000 size=0x08e30 ( 36400) load
I (280) boot: Loaded app from partition at offset 0x10000

Starting ESP8266 CPU with ID: 00199647
```

Note that the partition table is not the best, FWs are better if they start at 0x010000 and 0x110000 so you can use the same fw at either slot, but that's not working due to #16402 (larger partition tables work fine with a newer bootloader binary).

**Review by fjmolinas — APPROVED** [review]
ACK, CI is happy changes look good and @iosabi provided test output for the new functionality.
